week_2_typescript/problem_10.ts

```diff
@@ -0,0 +1,173 @@
+/*
+
+Intro:
+
+    We have asynchronous functions now, advanced technology.
+    This makes us a tech startup officially now.
+    But one of the consultants spoiled our dreams about
+    inevitable future IT leadership.
+    He said that callback-based asynchronicity is not
+    popular anymore and everyone should use Promises.
+    He promised that if we switch to Promises, this would
+    bring promising results.
+
+Exercise:
+
+    We don't want to reimplement all the data-requesting
+    functions. Let's decorate the old callback-based
+    functions with the new Promise-compatible result.
+    The final function should return a Promise which
+    would resolve with the final data directly
+    (i.e. users or admins) or would reject with an error
+    (or type Error).
+
+    The function should be named promisify.
+
+Higher difficulty bonus exercise:
+
+    Create a function promisifyAll which accepts an object
+    with functions and returns a new object where each of
+    the function is promisified.
+
+    Rewrite api creation accordingly:
+
+        const api = promisifyAll(oldApi);
+
+*/
+
+interface User {
+  type: "user";
+  name: string;
+  age: number;
+  occupation: string;
+}
+
+interface Admin {
+  type: "admin";
+  name: string;
+  age: number;
+  role: string;
+}
+
+type Person = User | Admin;
+
+const admins: Admin[] = [
+  { type: "admin", name: "Jane Doe", age: 32, role: "Administrator" },
+  { type: "admin", name: "Bruce Willis", age: 64, role: "World saver" },
+];
+
+const users: User[] = [
+  {
+    type: "user",
+    name: "Max Mustermann",
+    age: 25,
+    occupation: "Chimney sweep",
+  },
+  { type: "user", name: "Kate Müller", age: 23, occupation: "Astronaut" },
+];
+
+export type ApiResponse<T> =
+  | {
+      status: "success";
+      data: T;
+    }
+  | {
+      status: "error";
+      error: string;
+    };
+
+type callbackFunction<T> = (
+  callback: (response: ApiResponse<T>) => void
+) => void;
+type newPromise<T> = () => Promise<T>;
+export function promisify<T>(arg: callbackFunction<T>): newPromise<T> {
```

Comment:
Provide line breaks and spaces where-ever necessary to increase code readability.